client/src/cordova/android/OutlineAndroidLib/outline/src/main/java/org/outline/vpn/VpnTunnelService.java

```diff
@@ -272,19 +327,40 @@ private synchronized boolean isTunnelActive(final String tunnelId) {
     return tunnelConfig.id.equals(tunnelId);
   }
 
-  /* Helper method to tear down an active tunnel. */
+  /** Helper method to tear down an active tunnel. */
   private void tearDownActiveTunnel() {
     stopVpnTunnel();
     stopForeground();
-    tunnelConfig = null;
+    this.tunnelConfig = null;
     stopNetworkConnectivityMonitor();
-    tunnelStore.setTunnelStatus(TunnelStatus.DISCONNECTED);
+    this.tunnelStore.setTunnelStatus(TunnelStatus.DISCONNECTED);
   }
 
-  /* Helper method that stops the Outline client, tun2socks, and tears down the VPN. */
+  /** Helper method that stops the Outline client, tun2socks, and tears down the VPN. */
   private void stopVpnTunnel() {
-    vpnTunnel.disconnectTunnel();
-    vpnTunnel.tearDownVpn();
+    this.stopRemoteDevice();
+
+    if (this.tunFd != null) {
+      try {
+        // On close, the OS restores the routing and destroys the TUN device.
+        this.tunFd.close();
+      } catch (IOException e) {
+        LOG.severe("Failed to close the VPN interface file descriptor.");
+      } finally {
+        this.tunFd = null;
+      }
+    }
+  }
+
+  /** Stops the traffic exchange with the remote device. */
+  private synchronized void stopRemoteDevice() {
+    if (this.remoteDevice == null) {
+      return;
+    }
+    if (this.remoteDevice.isConnected()) {
+      this.remoteDevice.disconnect();
```

Comment:
Does `disconnect` have any side effects if connected? Maybe we can just call `disconnect` without checking `isConnected`?

Reply:
Apparently it doesn't. But I think it's not great practice to call disconnect twice or it it's not needed.